### PR TITLE
Add a way to configure the containerd namespace…

### PR DESCRIPTION
**Pull request by vdemeester**
This adds an option (and a flag) to be able to set a different
containerd namespace than the default one ("moby") to be used by
moby.

This should make it pretty easy to run multiple docker daemon pointing
to the same containerd without conflicting with each other. This also
change the plugin namespace.

:lion: 

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by cpuguy83** [comment]
There is a problem with this, the containerd runtime data is so deeply nested (when using Docker's containerd) that these namespace names have to be short otherwise we run into socket lenght issues.

We also would still need namespace for images (free with the c8d move), networks, and volumes.

**Comment by vdemeester** [comment]
> There is a problem with this, the containerd runtime data is so deeply nested (when using Docker's containerd) that these namespace names have to be short otherwise we run into socket lenght issues.

Oh right, didn't really think of that. 

> We also would still need namespace for images (free with the c8d move), networks, and volumes.

Definitely, but right now, they "can" be namespace by changing the data dir of `dockerd` right ? 

The main usage I "had" with that, is, when `containerd` is not managed by docker, to be able to start multiple daemon pointing to the same containerd :angel:

**Comment by mlaventure** [comment]
Doesn't build ;-):
```
10:34:31 plugin/executor/containerd/containerd.go:47:2: undefined: pluginNamespace
10:34:31 plugin/executor/containerd/containerd.go:48:34: undefined: pluginNamespace
```

Otherwise, since we currently don't use c8d atm for the images, this works for me. 

`containerd` should return a proper warning when the path is too long for the socket now normally.

**Comment by cpuguy83** [comment]
This seems OK.

**Comment by tonistiigi** [comment]
> This should make it pretty easy to run multiple docker daemon pointing
to the same containerd without conflicting with each other. This also
change the plugin namespace.

Curious, what is conflicting atm when using the same namespace? I remember this used to work fine in `v0.2`. (Not against making it configurable.)

**Comment by cpuguy83** [comment]
It simplifies CI a lot at least (granted there are other issues with
multi-daemon).

**Comment by anusha-ragunathan** [comment]
LGTM

**Review by anusha-ragunathan — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Oh, can you:

- Update the API version
- Add a note to the API changelog; https://github.com/moby/moby/blob/master/docs/api/version-history.md ?

File: daemon/daemon.go [hunk 1]
```diff
@@ -623,6 +623,10 @@ func NewDaemon(config *config.Config, registryService registry.Service, containe
 		os.Setenv("TMPDIR", realTmp)
 	}
 
+	if config.ContainerdNamespace == "" {
+		config.ContainerdNamespace = ContainersNamespace
```
Comment (thaJeztah) [inline daemon/daemon.go]: Looks like we need some validation on this;

```bash
dockerd --containerd-namespace="hello world"
```

```
DEBU[2018-05-19T19:06:41.987274386Z] processing event stream                       module=libcontainerd namespace="hello world"
ERRO[2018-05-19T19:06:41.987788322Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.987921624Z] processing event stream                       module=libcontainerd namespace="plugins.hello world"
ERRO[2018-05-19T19:06:41.988186851Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==plugins.hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="plugins.hello world"
DEBU[2018-05-19T19:06:41.988142306Z] processing event stream                       module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.988608030Z] processing event stream                       module=libcontainerd namespace="plugins.hello world"
ERRO[2018-05-19T19:06:41.988938267Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.989253238Z] processing event stream                       module=libcontainerd namespace="hello world"
ERRO[2018-05-19T19:06:41.989891739Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.990024270Z] processing event stream                       module=libcontainerd namespace="hello world"
ERRO[2018-05-19T19:06:41.989891331Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==plugins.hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="plugins.hello world"
DEBU[2018-05-19T19:06:41.990252468Z] processing event stream                       module=libcontainerd namespace="plugins.hello world"
ERRO[2018-05-19T19:06:41.990646964Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==plugins.hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="plugins.hello world"
DEBU[2018-05-19T19:06:41.990748019Z] processing event stream                       module=libcontainerd namespace="plugins.hello world"
ERRO[2018-05-19T19:06:41.991305511Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.991406990Z] processing event stream                       module=libcontainerd namespace="hello world"
ERRO[2018-05-19T19:06:41.991860574Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.992128312Z] processing event stream                       module=libcontainerd namespace="hello world"
ERRO[2018-05-19T19:06:41.991378514Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==plugins.hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="plugins.hello world"
ERRO[2018-05-19T19:06:41.992490201Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.992811583Z] processing event stream                       module=libcontainerd namespace="hello world"
^CERRO[2018-05-19T19:06:41.993449302Z] failed to get event                           error="rpc error: code = Unknown desc = subscription error: failed parsing subscription filters: filters: parse error: [namespace==hello >| |< world,topic~=|^/tasks/|]: unexpected input:  : invalid argument" module=libcontainerd namespace="hello world"
DEBU[2018-05-19T19:06:41.993673609Z] processing event stream                       module=libcontainerd namespace="hello world"
INFO[2018-05-19T19:06:41.993796788Z] Processing signal 'interrupt'                
DEBU[2018-05-19T19:06:41.994040940Z] daemon configured with a 15 seconds minimum shutdown timeout 
DEBU[2018-05-19T19:06:41.994092214Z] start clean shutdown of all containers with a 15 seconds timeout... 
```

However; this starts:

```bash
dockerd --containerd-namespace="test,namespace==foo"
```

But then running a container succeeds, but produces an error;;

```
docker: Error response from daemon: namespace validation: namespace "test,namespace==foo" must match ^[A-Za-z][A-Za-z0-9]+(:?[-]+[A-Za-z][A-Za-z0-9]+)*(?:[.](?:[A-Za-z][A-Za-z0-9]+(:?[-]+[A-Za-z][A-Za-z0-9]+)*))*$: invalid argument.
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Sorry, founds some things that still need to be updated 😅

**Comment by vdemeester** [comment]
@thaJeztah updated with validation (from `containerd` code :wink:)

```
$ dockerd --containerd-namespace="hello world"
namespace "hello world" must match ^[A-Za-z][A-Za-z0-9]+(:?[-]+[A-Za-z][A-Za-z0-9]+)*(?:[.](?:[A-Za-z][A-Za-z0-9]+(:?[-]
+[A-Za-z][A-Za-z0-9]+)*))*$: invalid argument  
```

File: daemon/daemon.go (hunk 1 above)
Comment (vdemeester) [reply daemon/daemon.go]: Validation happens here : https://github.com/moby/moby/pull/36914/files#diff-4eeac9ba6fec6b0677037c6f8735d16bR555

File: api/swagger.yaml [hunk 2]
```diff
@@ -3420,6 +3420,10 @@ definitions:
           > should not be considered stable.
         type: "string"
         example: "7TRN:IPZB:QYBB:VPBQ:UMPP:KARE:6ZNR:XE6T:7EWV:PKF4:ZOJD:TPYS"
+      ContainerdNamespace:
```
Comment (cpuguy83) [inline api/swagger.yaml]: Should this be a top-level key or something more opaque like a `map[string]string` for runtime config?

Comment (vdemeester) [reply api/swagger.yaml]: Hum good question, @thaJeztah ? :stuck_out_tongue:

**Comment by vdemeester** [comment]
ping @thaJeztah not sure what to do :stuck_out_tongue_closed_eyes:

**Comment by Rid** [comment]
Why was this closed @vdemeester? Are there any plans to allow running containers in different namespaces?

We are running a multi-tenant system and would like to separate containers into different namespaces to avoid conflicts.

**Pull request closed without merging**